tests/test.rs

```diff
@@ -0,0 +1,100 @@
+use insta;
+use os_pipe::pipe;
+use std::{io::Read, process::Command, sync::Mutex};
+use structopt::lazy_static::lazy_static;
+
+lazy_static! {
+    /// rust will try to run the tests in parallel by default, and `insta` doesn't like the
+    /// usual way of disabling this via `--test-threads=1`, so we're using this
+    /// mutex to make sure we're not re-flashing until the last run is finished
+    static ref ONE_RUN_AT_A_TIME: Mutex<i32> = Mutex::new(0i32);
+}
+
+/// run probe-run with `args` and truncate the "Finished .. in .." and "Running `...`" flashing output
+/// NOTE: this currently only capures `stdin`, so any `log::` ed output, like flashing
+fn run_and_truncate(args: &str) -> String {
+    let _guard = ONE_RUN_AT_A_TIME.lock().unwrap();
+
+    // add prefix to run this repository's version of `probe-run` and
+    // remove user-dependent registry and rustc information from backtrace paths
+    let complete_command = format!("run -- {} --shorten-paths", args);
+    let complete_command_args = complete_command.split(" ");
+
+    let (mut reader, writer) = pipe().unwrap();
+    let writer_clone = writer.try_clone().unwrap();
+
+    let mut handle = Command::new("cargo")
+        .args(complete_command_args)
+        // capture stderr and stdout while preserving line order
+        .stdout(writer)
+        .stderr(writer_clone)
+        // run `probe-run`
+        .spawn()
+        .unwrap();
+
+    // retrieve output and clean up
+    let mut probe_run_output = String::new();
+    reader.read_to_string(&mut probe_run_output).unwrap();
```

Comment:
the order of these operations seemed a bit unusual: I would wait for the process to finish and then read the pipe but here it's done the other way -- I guess `read_to_string` blocks until the two `writer` handles are closed so you don't see "incomplete data"?

out of curiosity can you do `status` (to drive the program to completion) instead of `spawn` and then read out from the `reader`? or does that not work due to how pipes work (it may be an error to read from the `reader` end when the `writer` end has been closed?)

Reply:
if we use `status()` (without `wait()` in because there's no child to wait for) the tests turn flaky; the command hangs in some cases but not all and it's not reproducible which and how many will pass. The same happens if we move `handle.wait()` before the string reading